### ISSUE TITLE
directives don't get processed

### ISSUE THREAD

**Issue by joaofernandes** [pbastowski/angular2-now#9]
Hi, I'm trying to get Angular 1 directives to get processed with angular2now but I'm unable to get them running.

My directive is defined as this 

var {Directive, Inject } = angular2now;
angular.module('App');
@Directive({
    selector:'header',
    restrict: 'E',
    scope: {},
    templateUrl: 'header-template.html',
    injectables: ['$compile', '$location']
})
export class Header
{
    constructor($compile, $location)
    {
...
    }

```
link(scope, element, attributes)
{  
}
```

}
In my controller 

import {Header} from 'header.js';
@View({
    templateUrl:'home.html',
    directives: [Header]
})
@Inject(['$scope','$location'])
@Controller('homeController')

and my header.html does use
&lt;header&gt;&lt;/header&gt;

but it's content is not processed.

Any hint on how to achieve it or how to debug it?

**Comment by joaofernandes** [comment]
Ok it seems that templateUrl gets overridden by target.templateUrl, is there any reason why? If I define in @Directive({templateUrl:.....}) shouldn't it be kept?

I've commented   "options.templateUrl = target.templateUrl || undefined;" and it stars working.

**Comment by pbastowski** [comment]
@joaofernandes Looking at your brief examples, I can see that you are not actually using the library the way it is intended, which is why you are not getting the results you expect. I would suggest reading the README again to familiarise yourself with the API. 

Also, you might want to have a look at my other example's on github, such as:
https://github.com/pbastowski/Thinkster-MEAN-Tutorial-in-angular-meteor/tree/feature/ng2-now-with-services
and
https://github.com/pbastowski/meteor-angular-socially/tree/feature/ng2now
and
https://github.com/pbastowski/todo-ng2now

Now, allow me to quickly show you how I would write your code and perhaps it will help.

``` javascript
var {SetModule, Component, Inject } = angular2now;

SetModule('App');  // SetModule will replace angular.module in angular2-now in the future

@Component({  selector: 'header' })
@View({ templateUrl: 'header-template.html' })
@Inject([ '$compile', '$location' ])

export class Header {
  constructor($compile, $location) {
    ...
  }
}
```

Notice, no `restrict` or `scope` is required. angular2-now always uses private scope. Do not use`Directives`, as it is not implemented, Don't use the link function, don't even write it in the code, a it is not necessary in most cases. The controller (constructor) has access to almost everything - you can inject `$element`, `$attrs` and `$scope` (if you really need it) into the constructor.

And I'm not sure what you are referring to as your "Controller". We don't use controllers in angular2-now. We create components, which can include other components, can have routes assigned to them, etc. No controllers necessary, except in very specialised cases. 

``` javascript
@View({ templateUrl: 'home.html' })
@Inject([ '$scope', '$location' ])
class homeController {
  constructor($scope, $location) {
  }
}
```

Notice that in angular2-now you don't import the Header class itself. You just use the `<header></header>` component in some html file where you need it.

Also, have you bootstrapped the application using code similar to this:

``` javascript
var {bootstrap, ... } = angular2now;

class myApplication {
  ...
}

bootstrap(myApplication)
```

In summary:
- read the angular2-now README and the examples again
- read the code in the two repositories I quote above

Good luck!

**Comment by joaofernandes** [comment]
Thanks for the reply, currently I don't have access to the code but I didn't use the bootstrap method because we're using AngularAMD to lazy load views/controllers only when required. Is this required to make our old directives work in this new format?

Regarding the $element, $attrib and $scope I'll try that tomorrow once I get to work, thanks for the tip.

Regarding the controller is the same approach as you have in your https://github.com/pbastowski/meteor-angular-socially/blob/step_17/client/routes.js defined in your stateProvider. How does your code know that it needs to download PartyDetailsCtrl file? I don't see any reference to it. Are you merging all your js to a single source? Is that why Components aren't referenced anywhere?

When we started this new app, we started with es6 transpiled to es5 with traceur and this new lib will help us to make it even more seamless the transition to angular 2.

**Comment by pbastowski** [comment]
I see the problem now: you are not looking at my angular2-now branch of socially, but the "older" style that uses controllers, etc.

Try this branch: https://github.com/pbastowski/meteor-angular-socially/tree/feature/ng2now.

And specifically, this shows you how to bootstrap:

https://github.com/pbastowski/meteor-angular-socially/blob/feature/ng2now/client/app/app.es6.js

Also, if you lazy-load angular modules after angular has bootstrapped, you won't be able to create any more components, directives, filters, services, etc. Angular does not support lazy loading of modules or directives natively at all. There are hacks and methods that allow you to create components and services after angular has been bootstrapped, but not straight out of the box.

My apps have all the code (js, html templates) bundled together into one js file. No lazy loading at all. In the examples that you are looking at, it is Meteor that does the continuous integration and the bundling for me.

Please note that I use Babel, not Traceur, for transpiling ES6 to E5. Thus, I don't know if angular2-now works properly with Traceur or what bugs you might encounter.

**Comment by joaofernandes** [comment]
Ok I got it working with both lazy loading and following your directions. My directives do get access to $scope,$element,$attrs with no problem and so, no link function is required.

My only issue right now is that my variables defined using bind

@Component({ selector:'tile',
    bind:
    {
            item: '=',
            template: '@',
            controller: '='
    }
})

are now stored in the controller and not in the $scope as before. I know that $scope is going away but if bindToController is true (default for angular2now), shouldn't I be able to use in my templates
{{ item }} instead of {{ tile.item }} ?

Btw, I'm also using babel now and it works the same way as traceur, it't only bit picky regarding constructors and super() which I can live with.

**Comment by pbastowski** [comment]
> if bindToController is true (default for angular2now), shouldn't I be able to use in my templates
> {{ item }} instead of {{ tile.item }} ?

Unfortunately no. bindToController still requires controllerAs, and you **can't** assign an empty string to controllerAs (controllerAs: ''). However, you could use a short prefix, such as 'vm', for example. Personally, I would also prefer to use no prefix, since each component get's it's own private scope, but there is no such option in AngularJS 1.4.

You can tell angular2now to always use 'vm' or any other string as your `controllerAs` by putting this code before you use any angular2now annotations.

``` javascript
angular2now.options({ controllerAs: 'vm' });
```

**Comment by pbastowski** [comment]
I'm closing this issue. It is not an angular2-now bug :)

**Comment by joaofernandes** [comment]
Hi thanks for the hints, I managed to get everything working as expected. It's really too bad requiring the controller reference it the views but it will be easy to migrate once 2.0 ships thanks to the global controller definition.

I noticed that Angular 2 does provide support for property selectors by defining [selector] instead of just selector. Do you think angular2now could support that instead of forcing AE for restrict for all directives?

**Comment by pbastowski** [comment]
You are welcome :)

More Angular 2 features are on my roadmap, but I can't say when I'll implement any of them.

For the moment, the library is working quite well for what myself and my team need from it. New features get implemented as required, when we identify a use case that reduces our coding, for instance, or improves maintainability.
